Migrate content from v5 (#2043)
plone/documentation: docs/backend/search.md

    "keywords": "Plone, search, catalog, query, brains, ZCatalog, portal_catalog, FieldIndex, KeywordIndex, DateIndex, searchResults"
---

(backend-search-label)=

# Search

To index and search content in Plone, see the Plone 5 documentation {doc}`plone5:develop/plone/searching_and_indexing/index`.
Searching is the action of retrieving data from Plone's catalog based on specific criteria.
In Plone, this typically means querying content items using either the `plone.api.content.find` function or directly using the `portal_catalog` tool.

Alternatively, you can integrate any open source search engine with your Plone site.
This chapter focuses on **querying** the catalog to find content.
For information about **indexing** content and making it searchable, see {doc}`/backend/indexing`.

-   [Solr](https://solr.apache.org/) - See the add-on [`collective.solr`](https://github.com/collective/collective.solr) and its [documentation](https://collectivesolr.readthedocs.io/en/latest/).
-   [`collective.elasticsearch`](https://github.com/collective/collective.elasticsearch)
-   [`collective.elastic.plone`](https://github.com/collective/collective.elastic.plone)

> You can find a comprehensive list of search options in [Awesome Plone - Searching and Categorizing](https://github.com/collective/awesome-plone?tab=readme-ov-file#searching-and-categorizing)

Reviewer: It would be good to keep this section about external search integrations.

(backend-search-indexes-vs-metadata-label)=

### Indexes versus metadata

The catalog manages two types of data:

Indexes
:   Searchable fields that you can query against.
    Different index types support different query operations.
    For example, `FieldIndex` for exact matches, `KeywordIndex` for list values, and `ZCTextIndex` for full-text search.

Metadata (columns)
:   Copies of object attributes stored in the catalog.
    Metadata values are returned with search results, allowing you to access common attributes without loading the full object.

You can view the available indexes and metadata columns through the Zope Management Interface (ZMI) by navigating to `portal_catalog`.

(backend-search-brains-label)=

### Catalog brains

When you search the catalog, the results are not actual content objects.
Instead, you receive *catalog brains*, which are lightweight proxy objects.

Brains are lazy in two ways:

1.  They are created only when your code requests each result.
2.  They don't load the actual content objects from the database.

This lazy behavior provides significant performance benefits.

> You can iterate through thousands of search results without loading any objects into memory.

Reviewer: ```suggestion
You can iterate through thousands of search results without loading full objects into memory.
```

Brains provide access to:

-   All metadata columns defined in the catalog
-   Methods to retrieve the actual object when needed
-   The path and URL of the indexed content

```{note}
Calling `brain.getObject()` loads the full object from the database.
This has performance implications when working with many results.
Use metadata whenever possible to avoid unnecessary database access.
```

(backend-search-other-catalogs-label)=

### Other catalogs

Besides `portal_catalog`, Plone maintains additional specialized catalogs:

`uid_catalog`
:   Maintains a lookup table for objects by their Unique Identifier (UID).
    UIDs remain constant even when objects are moved.

`reference_catalog`
:   Tracks inter-object references by UID.
    Used internally by relation fields.

> Add-on products may install their own catalogs optimized for specific purposes.

Reviewer: These catalogs were used in Archetypes which is obsolete in Plone 5.2. Remove this section please.

(backend-search-querying-label)=

## Querying the catalog

(backend-search-accessing-catalog-label)=

### Accessing the catalog

The recommended way to search for content is using `plone.api`:

```python
from plone import api

# Search using api.content.find (recommended)
results = api.content.find(portal_type='Document')

# Get the catalog tool directly
catalog = api.portal.get_tool('portal_catalog')
```

> You can also use the traditional `getToolByName` helper:

Reviewer: ```suggestion
Code in Plone core which cannot use `plone.api` should use the `getToolByName` helper instead:
```

```python
from Products.CMFCore.utils import getToolByName

catalog = getToolByName(context, 'portal_catalog')
```

(backend-search-performing-queries-label)=

### Performing queries

> There are several ways to query the catalog:

Reviewer: We should have a section that explains the difference between restricted and unrestricted queries. (Restricted queries automatically filter for content that can be viewed by the current user and is within its effective date range. Unrestricted queries don't include those filters.)

```python
from plone import api

# Using api.content.find (recommended for most cases)
results = api.content.find(portal_type='Document')

# Using the catalog directly
catalog = api.portal.get_tool('portal_catalog')
results = catalog(portal_type='Document')

# Using searchResults explicitly
results = catalog.searchResults(portal_type='Document')
```

Pass search criteria as keyword arguments, where the key is an index name and the value is the search term.

```python
# Find all published News Items
results = api.content.find(
    portal_type='News Item',
    review_state='published'
)
```

Multiple criteria are combined with logical AND.
The query above finds items that are both a News Item AND in the published state.

> Calling the catalog without arguments returns all indexed content:

Reviewer: This is out of date. In current versions of Plone it returns no results. To get all brains, use catalog.getAllBrains()

```python
# Get all content (use with caution on large sites)
all_brains = catalog()
```

(backend-search-working-with-brains-label)=

### Working with catalog brains

Search results are iterable collections of brain objects:

```python
from plone import api

results = api.content.find(portal_type='Document')

for brain in results:
    # Access metadata directly as attributes
    print(f"Title: {brain.Title}")
    print(f"Description: {brain.Description}")
    print(f"URL: {brain.getURL()}")
    print(f"Path: {brain.getPath()}")
```

To get the actual content object from a brain:

```python
# Load the full object (performance cost)
obj = brain.getObject()
```

```{warning}
Calling `getObject()` on many brains can significantly impact performance.

> Each call requires a separate database query.

Reviewer: This is misleading. Loading catalog brains also queries the database. But catalog brains are smaller and are loaded in batches.

Prefer using metadata columns when possible.
```

Common brain methods and attributes:

`brain.getObject()`
:   Returns the actual content object.
    Use sparingly due to performance cost.

`brain.getURL()`
:   Returns the absolute URL of the object.
    Equivalent to `obj.absolute_url()`.

`brain.getPath()`
:   Returns the physical path as a string.
    Equivalent to `'/'.join(obj.getPhysicalPath())`.

`brain.getRID()`

> :   Returns the internal result ID used by the catalog.

Reviewer: ```suggestion
:   Returns the internal record ID used by the catalog.
```
